pkgs/build-support/rust/fetchCargoTarball.nix

```diff
@@ -0,0 +1,81 @@
+{ stdenv, cacert, git, cargo, python3 }:
+let cargo-vendor-normalise = stdenv.mkDerivation {
+  name = "cargo-vendor-normalise";
+  src = ./cargo-vendor-normalise.py;
+  nativeBuildInputs = [ python3.pkgs.wrapPython ];
+  dontUnpack = true;
+  installPhase = "install -D $src $out/bin/cargo-vendor-normalise";
+  pythonPath = [ python3.pkgs.toml ];
+  postFixup = "wrapPythonPrograms";
+  doInstallCheck = true;
+  installCheckPhase = ''
+    # check that ./fetchcargo-default-config.toml is a fix point
+    reference=${./fetchcargo-default-config.toml}
+    < $reference $out/bin/cargo-vendor-normalise > test;
+    cmp test $reference
+  '';
+  preferLocalBuild = true;
+};
+in
+{ name ? "cargo-deps"
+, src ? null
+, srcs ? []
+, patches ? []
+, sourceRoot
+, sha256
+, cargoUpdateHook ? ""
+, ...
+} @ args:
+stdenv.mkDerivation ({
+  name = "${name}-vendor.tar.gz";
+  nativeBuildInputs = [ cacert git cargo-vendor-normalise cargo ];
+
+  phases = "unpackPhase patchPhase buildPhase installPhase";
+
+  buildPhase = ''
+    # Ensure deterministic Cargo vendor builds
+    export SOURCE_DATE_EPOCH=1
+
+    if [[ ! -f Cargo.lock ]]; then
+        echo
+        echo "ERROR: The Cargo.lock file doesn't exist"
+        echo
+        echo "Cargo.lock is needed to make sure that cargoSha256 doesn't change"
+        echo "when the registry is updated."
+        echo
+
+        exit 1
+    fi
+
+    # Keep the original around for copyLockfile
+    cp Cargo.lock Cargo.lock.orig
+
+    export CARGO_HOME=$(mktemp -d cargo-home.XXX)
+    CARGO_CONFIG=$(mktemp cargo-config.XXXX)
+
+    ${cargoUpdateHook}
+
+    cargo vendor $name | cargo-vendor-normalise > $CARGO_CONFIG
+
+    # Add the Cargo.lock to allow hash invalidation
+    cp Cargo.lock.orig $name/Cargo.lock
+
+    # Packages with git dependencies generate non-default cargo configs, so
+    # always install it rather than trying to write a standard default template.
+    install -D $CARGO_CONFIG $name/.cargo/config;
+  '';
+
+  # Build a reproducible tar, per instructions at https://reproducible-builds.org/docs/archives/
+  installPhase = ''
+    tar --owner=0 --group=0 --numeric-owner --format=gnu \
+        --sort=name --mtime="@$SOURCE_DATE_EPOCH" \
+        -czvf $out $name
```

Comment:
I'd remove `v` flag.

Reply:
Removed.  BTW, here's a concrete example of the difference that this makes to the `attr.cargoDeps` size:

```
# Ripgrep's cargoDeps as currently built on master:
156M    /nix/store/vdaa161ixsyjmvqpyf101nvkrf9w0ddm-ripgrep-11.0.2-vendor

# Ripgrep's cargoDeps if we were to set legacyCargoFetcher = false on this PR and update hash:
16M     /nix/store/cwafl8f9gradpyhs4mjygnkl8mijq558-ripgrep-11.0.2-vendor.tar.gz
```
156M -> 16M is a pretty big difference when you add up across all the Rust packages!

Reply:
I don't think it's worth spending too much CPU on this. The NAR files are already compressed (with xz) and the end-user installing ripgrep won't be downloading this.

Reply:
Sounds good, I'll leave it as is then.  

This does cause a rebuild of all the Rust applications, due to the attributes shuffling around, but no `cargoSha256`s or versions are materially changed, so I've currently got it going against `master`, but I can rebase and retarget for `staging` if we prefer.

Reply:
Yeah I think this is due for staging due to the amount of rebuild. It might also make sense to do the whole migration as a single commit to reduce the amount of rebuild.

Reply:
I plan to move relatively quickly on this once it lands on master, but ideally would like to proceed in the incremental steps outlined in the doc to attempt to reduce merge conflicts.

Already I had to carry 1 commit over from master to resolve a merge conflict for the git-workspace tool :)